Source/WebCore/dom/Document.cpp

```diff
@@ -4965,6 +4971,66 @@ void Document::noteUserInteractionWithMediaElement()
     updateIsPlayingMedia();
 }
 
+#if ENABLE(MEDIA_STREAM)
+static MediaProducerMediaStateFlags captureSourceState(const RealtimeMediaSource& source)
+{
+    ASSERT(source.isCaptureSource());
+
+    if (source.isEnded())
+        return MediaProducer::IsNotPlaying;
+
+    switch (source.deviceType()) {
+    case CaptureDevice::DeviceType::Microphone:
+        if (source.muted())
+            return MediaProducerMediaState::HasMutedAudioCaptureDevice;
+        if (source.interrupted())
+            return MediaProducerMediaState::HasInterruptedAudioCaptureDevice;
+        if (source.isProducingData())
+            return MediaProducerMediaState::HasActiveAudioCaptureDevice;
+        break;
+    case CaptureDevice::DeviceType::Camera:
+        if (source.muted())
+            return MediaProducerMediaState::HasMutedVideoCaptureDevice;
+        if (source.interrupted())
+            return MediaProducerMediaState::HasInterruptedVideoCaptureDevice;
+        if (source.isProducingData())
+            return MediaProducerMediaState::HasActiveVideoCaptureDevice;
+        break;
+    case CaptureDevice::DeviceType::Screen:
+        if (source.muted())
+            return MediaProducerMediaState::HasMutedScreenCaptureDevice;
+        if (source.interrupted())
+            return MediaProducerMediaState::HasInterruptedScreenCaptureDevice;
+        if (source.isProducingData())
+            return MediaProducerMediaState::HasActiveScreenCaptureDevice;
+        break;
+    case CaptureDevice::DeviceType::Window:
+        if (source.muted())
+            return MediaProducerMediaState::HasMutedWindowCaptureDevice;
+        if (source.interrupted())
+            return MediaProducerMediaState::HasInterruptedWindowCaptureDevice;
+        if (source.isProducingData())
+            return MediaProducerMediaState::HasActiveWindowCaptureDevice;
+        break;
+    case CaptureDevice::DeviceType::SystemAudio:
+    case CaptureDevice::DeviceType::Speaker:
+    case CaptureDevice::DeviceType::Unknown:
+        ASSERT_NOT_REACHED();
+        break;
+    }
```

Comment:
that seems like a lot of duplicated code, couldn't we make the static method take a `CaptureDevice::DeviceType` instead and put this code in common?

Reply:
Will check what I can do there.

Reply:
Done by sharing code with MediaStreamTrack.